flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/row/ParquetRowDataWriter.java

```diff
@@ -224,6 +293,176 @@ private TimestampWriter(int precision) {
+                    Object key = keyElementGetter.getElementOrNull(keyArray, i);
+                    Object value = valueElementGetter.getElementOrNull(valueArray, i);
+
+                    recordConsumer.startGroup();
+                    if (key != null) {
+                        // write key element
+                        recordConsumer.startField(keyName, 0);
+                        keyWriter.write(key);
+                        recordConsumer.endField(keyName, 0);
+
+                        // write value element
+                        if (value != null) {
+                            recordConsumer.startField(valueName, 1);
+                            valueWriter.write(value);
+                            recordConsumer.endField(valueName, 1);
+                        }
+                    }
+                    recordConsumer.endGroup();
+                }
+
+                recordConsumer.endField(repeatedGroupName, 0);
+            }
+            recordConsumer.endGroup();
+        }
+
+        @Override
+        public void write(Object value) {}
+    }
+
+    private class ArrayWriter implements FieldWriter {
+
+        private String elementName;
+        private FieldWriter elementWriter;
+        private String repeatedGroupName;
+        private ArrayData.ElementGetter elementGetter;
+
+        private ArrayWriter(LogicalType t, GroupType groupType) {
+
+            // Get the internal array structure
+            GroupType repeatedType = groupType.getType(0).asGroupType();
+            this.repeatedGroupName = repeatedType.getName();
+
+            Type elementType = repeatedType.getType(0);
+            this.elementName = elementType.getName();
+
+            this.elementWriter = createWriter(t, elementType);
+            this.elementGetter = ArrayData.createElementGetter(t);
+        }
+
+        @Override
+        public void write(RowData row, int ordinal) {
+            recordConsumer.startGroup();
+            ArrayData arrayData = row.getArray(ordinal);
+            int listLength = arrayData.size();
+
+            if (listLength > 0) {
+                recordConsumer.startField(repeatedGroupName, 0);
+
+                for (int i = 0; i < listLength; i++) {
+                    Object object = elementGetter.getElementOrNull(arrayData, i);
+                    recordConsumer.startGroup();
+                    if (object != null) {
+                        recordConsumer.startField(elementName, 0);
+                        elementWriter.write(object);
+                        recordConsumer.endField(elementName, 0);
+                    }
+                    recordConsumer.endGroup();
+                }
+
+                recordConsumer.endField(repeatedGroupName, 0);
+            }
+            recordConsumer.endGroup();
+        }
+
+        @Override
+        public void write(Object value) {}
+    }
+
+    private class GroupTypeWriter implements FieldWriter {
```

Comment:
Why separate a class `GroupTypeWriter`? Merge this with `RowWriter`?

Reply:
merged